### PR TITLE
feat(spanner): add spanner_copy_backup sample and changes to spanner_update_ba…

### PR DESCRIPTION
**Pull request by rajatbhatta**
…ckup sample.

This commit incorporates the following changes:

-   Add sample for copy backup feature in spanner.

-   Make some changes to update backup feature in spanner to make sure
    expire time is always less than max expire set.

### CODE REVIEW

File: spanner/spanner_snippets/spanner/spanner_copy_backup.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"fmt"
+	"io"
+	"time"
+
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	pbt "github.com/golang/protobuf/ptypes/timestamp"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+)
+
+func copyBackup(ctx context.Context, w io.Writer, instanceId string, copyBackupId string, sourceBackupId string) error {
```
Comment (rahul2393) [inline spanner/spanner_snippets/spanner/spanner_copy_backup.go:30]: Could you please add comments for these functions: https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#comment-functions-and-packages

Comment (rahul2393) [reply spanner/spanner_snippets/spanner/spanner_copy_backup.go:30]: Please provide example values for functions: https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#function-arguments-for-snippets

Comment (rajatbhatta) [reply spanner/spanner_snippets/spanner/spanner_copy_backup.go:31]: These have been incorporated.

File: spanner/spanner_snippets/spanner/spanner_copy_backup.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"io"
+	"time"
+
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	pbt "github.com/golang/protobuf/ptypes/timestamp"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+)
+
+// copyBackup copies an existing backup to a given instance in same or different region, or in same or different project.
+func copyBackup(ctx context.Context, w io.Writer, instanceId string, copyBackupId string, sourceBackupId string) error {
```
Comment (codyoss) [inline spanner/spanner_snippets/spanner/spanner_copy_backup.go:31]: nit: [Declare a context.Context as needed](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#declare-a-contextcontext-as-needed)

**Review by codyoss — COMMENTED** [review]
Just a couple of things.

File: spanner/spanner_snippets/spanner/spanner_copy_backup.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	adminClient, err := database.NewDatabaseAdminClient(ctx)
+	if err != nil {
+		return fmt.Errorf("database.NewDatabaseAdminClient: %v", err)
+	}
+	defer adminClient.Close()
+
+	expireTime := time.Now().AddDate(0, 0, 14)
+
+	// Instantiate the request for performing copy backup operation.
+	copyBackupReq := adminpb.CopyBackupRequest{
```
[the remark this answers is not in the document]
Comment (rajatbhatta) [reply spanner/spanner_snippets/spanner/spanner_copy_backup.go:44]: Yes, right. Some changes are made to the proto, and the code is genproto will be autogenerated once proto changes are published. Which is why some builds are failing - including the above compile issue.

A newer version of genproto was generated locally, and tested against.

File: spanner/spanner_snippets/spanner/spanner_copy_backup.go (hunk 2 above)
Comment (rajatbhatta) [reply spanner/spanner_snippets/spanner/spanner_copy_backup.go]: Done.

**Comment by rajatbhatta** [comment]
@rahul2393 @codyoss : Can you please take a look at this PR? Some more changes have been made.

**Review by rahul2393 — APPROVED** [review]

File: spanner/spanner_snippets/spanner/spanner_update_backup.go [hunk 4]
```diff
@@ -24,12 +24,15 @@ import (
...
 	database "cloud.google.com/go/spanner/admin/database/apiv1"
-	pbts "github.com/golang/protobuf/ptypes"
 	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
 	"google.golang.org/genproto/protobuf/field_mask"
+	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
+// updateBackup updates the expiration time of a pending or completed backup.
 func updateBackup(ctx context.Context, w io.Writer, db, backupID string) error {
+	// db := "projects/my-project/instances/my-instance/databases/my-database"
```
Comment (codyoss) [inline spanner/spanner_snippets/spanner/spanner_update_backup.go:34]: Please declare context as needed.

**Review by codyoss — CHANGES_REQUESTED** [review]
Just a couple more small things

File: spanner/spanner_snippets/spanner/spanner_update_backup.go (hunk 4 above)
Comment (rajatbhatta) [reply spanner/spanner_snippets/spanner/spanner_update_backup.go:34]: This is done.

File: spanner/spanner_snippets/spanner/spanner_list_backup_operations.go [hunk 5]
```diff
@@ -28,7 +28,9 @@ import (
 	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
 )
 
-func listBackupOperations(ctx context.Context, w io.Writer, db string) error {
+func listBackupOperations(ctx context.Context, w io.Writer, db string, backupId string) error {
```
[the remark this answers is not in the document]
Comment (rajatbhatta) [reply spanner/spanner_snippets/spanner/spanner_list_backup_operations.go]: This is done.

**Comment by rajatbhatta** [comment]
@codyoss: I've made the changes based on review comments. Please take a look once if everything looks good.

File: spanner/spanner_snippets/spanner/spanner_list_backup_operations.go [hunk 6]
```diff
@@ -21,14 +21,23 @@ import (
...
 )
 
-func listBackupOperations(ctx context.Context, w io.Writer, db string) error {
+// listBackupOperations lists the backup operations that are pending or have completed/failed/cancelled within the last 7 days.
+func listBackupOperations(w io.Writer, db string, backupId string) error {
+	// db := "projects/my-project/instances/my-instance/databases/my-database"
+	// backupID := "my-backup"
+
+	// Add timeout to context.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Hour)
```
Comment (codyoss) [inline spanner/spanner_snippets/spanner/spanner_list_backup_operations.go:38]: Can it really take up an hour for a list operation? If not I think maybe just declare a Background context for the sample

Comment (rajatbhatta) [reply spanner/spanner_snippets/spanner/spanner_list_backup_operations.go:38]: It doesn't take that long. Declaring a Background context makes sense here.

Comment (rajatbhatta) [reply spanner/spanner_snippets/spanner/spanner_list_backup_operations.go]: This is done.

**Review by codyoss — APPROVED** [review]

**Review by rahul2393 — APPROVED** [review]

**Pull request merged**